deps/oculus/src/ovrsession.cpp

```diff
@@ -0,0 +1,456 @@
+    desc.Width = bufferSize.w;
+    desc.Height = bufferSize.h;
+    desc.MipLevels = 1;
+    desc.SampleCount = 1;
+    desc.StaticImage = ovrFalse;
+
+    result = ovr_CreateTextureSwapChainGL(session, &desc, &eyes[eye].ColorTextureChain);
+    int length = 0;
+    ovr_GetTextureSwapChainLength(session, eyes[eye].ColorTextureChain, &length);
+
+    if (!OVR_SUCCESS(result)) {
+      std::cout << "Error creating Oculus GL Swap Chain" << std::endl;
+    } else {
+      for (int i = 0; i < length; ++i) {
+        GLuint textureId;
+        ovr_GetTextureSwapChainBufferGL(session, eyes[eye].ColorTextureChain, i, &textureId);
+        glBindTexture(GL_TEXTURE_2D, textureId);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+      }
+    }
+
+    desc.Format = OVR_FORMAT_D32_FLOAT;
+
+    result = ovr_CreateTextureSwapChainGL(session, &desc, &eyes[eye].DepthTextureChain);
+    ovr_GetTextureSwapChainLength(session, eyes[eye].DepthTextureChain, &length);
+
+    if (!OVR_SUCCESS(result)) {
+    } else {
+      for (int i = 0; i < length; ++i) {
+        GLuint textureId;
+        ovr_GetTextureSwapChainBufferGL(session, eyes[eye].DepthTextureChain, i, &textureId);
+        glBindTexture(GL_TEXTURE_2D, textureId);
+
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+      }
+    }
+
+    glGenFramebuffers(1, fboId);
+
+  }
+}
+
+NAN_METHOD(OVRSession::GetRecommendedRenderTargetSize)
+{
+  if (info.Length() != 0)
+  {
+    Nan::ThrowError("Wrong number of arguments.");
+    return;
+  }
+
+  ovrSession session = *ObjectWrap::Unwrap<OVRSession>(info.Holder())->self_;
+  ovrHmdDesc hmdDesc = ObjectWrap::Unwrap<OVRSession>(info.Holder())->hmdDesc;
+
+  // Oculus: Initialize Swap Chain
+  ovrSizei leftEyeTextureSize = ovr_GetFovTextureSize(session, ovrEye_Left, hmdDesc.DefaultEyeFov[ovrEye_Left], 1);
+  ovrSizei rightEyeTextureSize = ovr_GetFovTextureSize(session, ovrEye_Right, hmdDesc.DefaultEyeFov[ovrEye_Right], 1);
+
+  int width  = leftEyeTextureSize.w;
+  int height = std::max(leftEyeTextureSize.h, rightEyeTextureSize.h);
+
+  Local<Object> result = Nan::New<Object>();
+  {
+    Local<String> width_prop = Nan::New<String>("width").ToLocalChecked();
+    Nan::Set(result, width_prop, Nan::New<Number>(width));
+
+    Local<String> height_prop = Nan::New<String>("height").ToLocalChecked();
+    Nan::Set(result, height_prop, Nan::New<Number>(height));
+  }
+  info.GetReturnValue().Set(result);
+}
+
+NAN_METHOD(OVRSession::GetPose) {
```

Comment:
I think this will need to be run in a thread soon -- either in this PR or it will block other PRs.

I think the `GetPose`/`Submit` flow is throttled by the oculus runtime to maintain framerate -- by design -- but this will probably lock up the node loop while it's happening, resulting in things like network not processing, or being reduced in throughput.

The strategy elsewhere is to do this work in a thread/Promise wrapping, while leaving the main loop to continue working while the runtime is waiting for the headset to flip framebuffers.

Also I think this is a logical conflict with the threading PR and might result in total lockups there (https://github.com/webmixedreality/exokit/pull/760).

Reply:
Here is the above strategy executed for the OpenVR case:

https://github.com/webmixedreality/exokit/blob/607d5a9512fe79df7a56551325065e62d24934a1/src/index.js#L1160

https://github.com/webmixedreality/exokit/blob/607d5a9512fe79df7a56551325065e62d24934a1/deps/openvr/src/ivrcompositor.cpp#L280